qemu/exec.c

```diff
@@ -1145,6 +1164,7 @@ void qemu_ram_free(struct uc_struct *uc, RAMBlock *block)
 
     QLIST_REMOVE_RCU(block, next);
     uc->ram_list.mru_block = NULL;
+    uc->ram_list.freed = true;
```

Comment:
I didn't understand why a new field `freed` was added.

Reply:
This is an optimization. With the benchmark we found that ram write time increase over the number of snapshots. So I have look at this with a profiler and found that most time is spend in `find_ram_offset()`. This function search for the best fit of a new `RAMBlock` in the `ram_addr_t` address space (special internal address space for RAMBlocks). Because of the data structure this takes O(n^2). 

When no `RAMBlock` was freed in this address space the new `RAMBlock` will be added at the end. `ram_list.freed` stores this information. Based on this boolean `find_ram_offset_last()` is called. This function skips the expensive search and just add the RAMBlock at the end of the address space.

Reply:
I read `find_ram_offset` again and think a bit. Is it possible that we disable this optimization for fragmentation?

```C
        /* If it fits remember our place and remember the size
         * of gap, but keep going so that we might find a smaller
         * gap to fill so avoiding fragmentation.
         */
        if (next - candidate >= size && next - candidate < mingap) {
            offset = candidate;
            mingap = next - candidate;
        }
````

In other words:

```C
        /* If it fits remember our place and remember the size
         * of gap, but keep going so that we might find a smaller
         * gap to fill so avoiding fragmentation.
         */
        if (next - candidate >= size && next - candidate < mingap) {
            offset = candidate;
            mingap = next - candidate;
            break; // Stop searching.
        }
```

Could you benchmark if this solution is fast enough?

Reply:
As far as I understand is the ram_addr_t address space an optimization for the dirty memory tracking. The allocator avoids fragmentation to have less bookkeeping in the memory tracking. So changing this allocator to fragment more might influence the execution speed in general, independent from the use of snapshots.

> Could you benchmark if this solution is fast enough?

Will do, but my benchmark only benchmark the snapshots, not the overall execution of an complex binary with dynamic maps/unmaps. So I can't tell if this change influence the emulation speed.

Reply:
I also guess having many ram blocks is quite uncommon. Also dynamic allocating many ram blocks is not a thing in qemu.